src/transformers/tools/agents.py

```diff
@@ -121,658 +122,369 @@ def _setup_default_tools():
-    while not lines[idx].lstrip().startswith("```"):
-        idx += 1
-    code = "\n".join(lines[start_idx:idx]).strip()
-
-    return explanation, code
-
-
-def clean_code_for_run(result):
-    result = f"I will use the following {result}"
-    explanation, code = result.split("Answer:")
-    explanation = explanation.strip()
+def clean_code_for_run(code):
     code = code.strip()
-
     code_lines = code.split("\n")
     if code_lines[0] in ["```", "```py", "```python"]:
         code_lines = code_lines[1:]
     if code_lines[-1] == "```":
         code_lines = code_lines[:-1]
     code = "\n".join(code_lines)
+    return code
+
+
+def parse_json_tool_call(json_blob: str):
+    try:
+        first_accolade_index =  json_blob.find("{")
+        last_accolade_index = [a.start() for a in list(re.finditer('}', json_blob))][-1]
+        json_blob = json_blob[first_accolade_index:last_accolade_index+1]
+        json_blob = literal_eval(json_blob)
+    except Exception as e:
+        raise ValueError(f"The JSON blob you used is invalid: due to the following error: {e}. Try to correct its formatting.")
+    if "action" in json_blob and "action_input" in json_blob:
+        return json_blob["action"], json_blob["action_input"]
+    else:
+        raise ValueError(f"Missing keys: {[key for key in ['action', 'action_input'] if key not in json_blob]} in blob {json_blob}")
+
+
+def format_prompt(toolbox, prompt_template,tool_description_template):
+    tool_descriptions = "\n".join([get_tool_description_with_args(tool, tool_description_template) for tool in toolbox.values()])
+    prompt = prompt_template.replace("<<tool_descriptions>>", tool_descriptions)
+    if "<<tool_names>>" in prompt:
+        tool_names = [f"'{tool_name}'" for tool_name in toolbox.keys()]
+        prompt = prompt.replace("<<tool_names>>", ", ".join(tool_names))
+    return prompt
+
+def to_text(input: Union[List[Dict[str, str]], Dict[str, str], str]) -> str:
+    if isinstance(input, list):
+        return "\n".join(map(lambda m: m["content"], input))
+    elif isinstance(input, dict):
+        return input["content"]
+    else:
+        return input
 
-    return explanation, code
+class Agent:
+    def __init__(
+            self, 
+            llm_callable,  
+            system_prompt=DEFAULT_REACT_SYSTEM_PROMPT, # TODO write default agent prompt
+            tool_description_template=None,
+            additional_args={},
+            max_iterations=1,
+            tool_parser=parse_json_tool_call,
+            toolbox=None, 
+        ):
+        
+        self.agent_name = self.__class__.__name__
+        self.llm_callable = llm_callable
+        self.prompt_template = system_prompt
+        self.tool_description_template = tool_description_template if tool_description_template else self.default_tool_description_template
+        self.additional_args = additional_args
+        self.max_iterations = max_iterations
+        self.log = lambda x: print(to_text(x))
```

Comment:
@LysandreJik thank you, we will use the transformer's `logging` mechanism as default.
We would need to log these:
    - agent rationale (can be done with `logger.info`)
    - bugs (`logger.debug`)
     - tool call
    - tool output
So tool calls and outputs should be logged differently: maybe we can enrich the default transformer's `logging` to do this?
On longer term we think of building a Gradio UI that would start on agent `run()` launch and lets you explore all steps taken by your agent. So this could be a drop-in replacement for the default`logger`.